security.py

```diff
@@ -64,8 +64,69 @@ def inner(request, *args, **kwargs):
             )
 
             if object_to_check.notified:
-                raise PermissionDenied("Notification for this assignment has already been sent.")
+                raise PermissionDenied(
+                    "Notification for this assignment has already been sent.",
+                )
 
             return func(request, *args, **kwargs)
         return inner
     return decorator
+
+
+def user_can_manage_file(func):
+    """
+    A decorator for checking if the current user can manage a file.
+    """
+    def wrapper(request, *args, **kwargs):
+        file_object_id = kwargs.get('file_id', None)
+
+        if not file_object_id:
+            raise Http404
+
+        file_object = get_object_or_404(
+            core_models.File,
+            pk=file_object_id,
+        )
+
+        if can_manage_file(request, file_object):
+            return func(request, *args, **kwargs)
+
+        return deny_access(request)
+
+    return wrapper
+
+
+
+def can_manage_file(request, file_object):
+    """
+    Determines if a user can view and download a file in the Typesetting Plugin.
+    """
+    if request.user.is_anonymous():
+        return redirect(
+            '{0}?next={1}'.format(
+                reverse('core_login'),
+                request.path_info
+            )
+        )
+
+    if file_object.article_id:
+        # Check if there is a workflow log entry for the typesetting plugin.
+        if not core_models.WorkflowLog.objects.filter(
+            article__pk=file_object.article_id,
+            element__element_name=plugin_settings.PLUGIN_NAME,
```

Comment:
worth checking `article__journal=request.journal` to avoid cross-journal URL forgery